File: new_LC_website_leethub2/0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        visit = set()
        rows, cols = len(grid), len(grid[0])
        res = 0
        max_area = 0
        def bfs(r,c):
            nonlocal res
            q = deque()
            visit.add((r,c))
            q.append((r,c))
            
            dirs = [[1,0],[-1,0],[0,1],[0,-1]]
            while q:
                row, col = q.popleft()
                for dr, dc in dirs:
                    r, c = row+dr, col+dc
                    if (r in range(rows) and
                    c in range(cols) and
                    grid[r][c] == 1 and
                    (r,c) not in visit):
                        res+=1
                        # print(res)
                        visit.add((r,c))
                        q.append((r,c))
            
        for r in range(0,rows):
            for c in range(0,cols):
                if grid[r][c] == 1 and (r,c) not in visit:
                    res = 1
                    visit.add((r,c))
                    bfs(r,c)
                    max_area = max(max_area,res)
        return max_area       

File: new_LC_website_leethub2/0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    int area = 0;
    
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int m = grid.size();
        
        int result = 0;
        int Ar = 0;
        for (int i = 0; i < m; i++) {
            int n = grid[i].size();
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 1) {
                    cout << "Grid["<<i<<"]["<<j<<"] "<<grid[i][j] << endl;
                    area = dfs(grid, i, j, m, n);
                    result++;
                }
                // cout << Ar << " "<<area << " "<<result<<endl;
                Ar = std::max(area,Ar);
                area = 0;
            }
        }
        
        return Ar;
    }
private:
    int dfs(vector<vector<int>>& grid, int i, int j, int m, int n) {
        
        // cout << area << " "<<i <<" "<< j <<endl;
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == 0) {
            return area;
        }
        area++;
        grid[i][j] = 0;
        
        dfs(grid, i - 1, j, m, n);
        dfs(grid, i + 1, j, m, n);
        dfs(grid, i, j - 1, m, n);
        dfs(grid, i, j + 1, m, n);
        
        return area;
    }
};